discovery/scaleway/scaleway.go

```diff
@@ -0,0 +1,213 @@
+
+	RefreshInterval model.Duration `yaml:"refresh_interval"`
+	Port            int            `yaml:"port"`
+}
+
+func (c SDConfig) Name() string {
+	return "scaleway"
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+func (c SDConfig) NewDiscoverer(options discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(&c, options.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// Discovery periodically performs Scaleway requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client    *scw.Client
+	port      int
+	zone      string
+	project   string
+	accessKey string
+	secretKey string
+	name      string
+	tags      []string
+}
+
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port:      conf.Port,
+		zone:      conf.Zone,
+		project:   conf.Project,
+		accessKey: conf.AccessKey,
+		secretKey: conf.SecretKey,
+		name:      conf.FilterName,
+		tags:      conf.FilterTags,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "scaleway_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	profile, err := d.loadProfile()
+	if err != nil {
+		return nil, err
+	}
+	d.client, err = scw.NewClient(
+		scw.WithHTTPClient(&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		}),
+		scw.WithUserAgent(fmt.Sprintf("Prometheus/%s", version.Version)),
+		scw.WithProfile(profile),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up scaleway client: %w", err)
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"scaleway",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "scaleway",
+	}
+
+	// instance fetching
+	instanceServerLabelSet, err := d.listInstanceServers(ctx)
+	if err != nil {
+		return nil, err
+	}
+	tg.Targets = append(tg.Targets, instanceServerLabelSet...)
+
+	// baremetal fetching
+	baremetalServerLabelSet, err := d.listBaremetalServers(ctx)
+	if err != nil {
+		return nil, err
+	}
+	tg.Targets = append(tg.Targets, baremetalServerLabelSet...)
+
+	return []*targetgroup.Group{tg}, nil
+}
+
+func (d *Discovery) loadProfile() (*scw.Profile, error) {
+	scwConfig, err := scw.LoadConfig()
+	// If the config file do not exist, don't return an error as we may find config in ENV or flags.
+	if _, isNotFoundError := err.(*scw.ConfigFileNotFoundError); isNotFoundError {
```

Comment:
At this stage we should have the information self-contained in the prometheus config.